pkg/api/customization/cluster/actions_rketemplate.go

```diff
@@ -0,0 +1,159 @@
+		return httperror.WrapAPIError(err, httperror.NotFound,
+			fmt.Sprintf("cluster with id %v doesn't exist", apiContext.ID))
+	}
+
+	if cluster.DeletionTimestamp != nil {
+		return httperror.NewAPIError(httperror.InvalidType,
+			fmt.Sprintf("cluster with id %v is being deleted", apiContext.ID))
+	}
+	if !v3.ClusterConditionReady.IsTrue(cluster) {
+		return httperror.WrapAPIError(err, httperror.ClusterUnavailable,
+			fmt.Sprintf("cluster not ready"))
+	}
+
+	//Copy the cluster rke spec to a template and revision object
+
+	actionInput, err := parse.ReadBody(apiContext.Request)
+	if err != nil {
+		return err
+	}
+	clusterTempName, ok := actionInput["clusterTemplateName"].(string)
+	if !ok || len(clusterTempName) == 0 {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, "must specify a name for the RKE Template")
+	}
+	clusterTempRevName, ok := actionInput["clusterTemplateRevisionName"].(string)
+	if !ok || len(clusterTempRevName) == 0 {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, "must specify a name for the RKE Template Revision")
+	}
+
+	creatorID := apiContext.Request.Header.Get("Impersonate-User")
+
+	newTemplateObj := newClusterTemplate(clusterTempName, cluster, creatorID)
+	clusterTemplate, err := a.ClusterTemplateClient.Create(newTemplateObj)
+	if err != nil {
+		return httperror.WrapAPIError(err, httperror.ServerError,
+			fmt.Sprintf("failed to create RKE Template object"))
+	}
+
+	newTemplateRevObj := newClusterTemplateRevision(clusterTempRevName, cluster, clusterTemplate, creatorID)
+	clusterTemplateRevision, err := a.ClusterTemplateRevisionClient.Create(newTemplateRevObj)
+	if err != nil {
+		//delete the clusterTemplate created above
+		deleteErr := a.ClusterTemplateClient.Delete(clusterTemplate.Name, &metav1.DeleteOptions{})
+		if deleteErr != nil {
+			logrus.Errorf("Failed to delete the RKE Template %v", clusterTemplate.Name)
+		}
+
+		return httperror.WrapAPIError(err, httperror.ServerError,
+			fmt.Sprintf("failed to create RKE Template Revision object"))
+	}
+
+	updatedCluster := cluster.DeepCopy()
+	updatedCluster.Spec.ClusterTemplateRevisionName = ref.Ref(clusterTemplateRevision)
+	updatedCluster.Spec.ClusterTemplateName = ref.Ref(clusterTemplate)
+
+	// Can't add either too many retries or longer interval as this an API handler
+	for i := 0; i < NumberOfRetriesForClusterUpdate; i++ {
+		_, err = a.ClusterClient.Update(updatedCluster)
+		if err == nil {
+			break
+		}
+		time.Sleep(RetryIntervalInMilliseconds * time.Millisecond)
+		cluster, err = a.ClusterClient.Get(apiContext.ID, v1.GetOptions{})
+		if err != nil {
+			logrus.Errorf("error fetching cluster with id %v: %v", apiContext.ID, err)
+			continue
+		}
+		updatedCluster = cluster.DeepCopy()
+		updatedCluster.Spec.ClusterTemplateRevisionName = ref.Ref(clusterTemplateRevision)
+		updatedCluster.Spec.ClusterTemplateName = ref.Ref(clusterTemplate)
+	}
+
+	apiContext.WriteResponse(http.StatusNoContent, map[string]interface{}{})
```

Comment:
should we return an explicit api error to the user if the update cluster with the template failed?

Reply:
done